### DISCUSSION TITLE
CDN Fallbacks?

### DISCUSSION THREAD

**Discussion by bryevdv** [bokeh/bokeh#12409] [Internals and design (Q&A)]
The cost and effort to maintain `cdn.bokeh.org` is nominal, and absent any unforeseen major problems, it should be sustainable indefinitely in to the future. Nevertheless, it's not *entirely* outside the realm of possibility that some problem with it could arise. Having it be a single point of failure, hard-coded into all usage does therefore incur at least some measure of operational risk. 

I had been thinking about the possibility of uploading BokehJS to a second CDN service that could be used as a fallback. I learned today that `jsdelivr` evidently automatically publishes everything uploaded to `npmjs` already, without any intervention. For instance:

https://cdn.jsdelivr.net/npm/@bokeh/bokehjs@2.4.3/build/js/bokeh.min.js

More generally, everything for the BokehJS project is viewable here: 

https://www.jsdelivr.com/package/npm/@bokeh/bokehjs

I think this greatly lowers the barrier to using an alternate public CDN in some way. I think there are various actions we could consider. 

* Add documentation  that references the availability of the `jsdelivr` option and how to use it
* Implement technical fallback mechanisms that will load from `jsdelivr` in the event the Bokeh CDN fails [1]

Other ideas? cc @bokeh/core for thoughts. 

[1] Some research indicates that this is possible, if clunky, even for script tags in plain HTML.

**Comment by philippjfr** [comment]
+1 

Adding easy mechanism to configure the CDN both for bokeh.js and its extensions (like panel) seems like a great idea. Both unpkg.com and jsdelivr are good options but since their sponsors could decide to stop supporting them they also carry operational risk so redundancy via fallbacks would be fabulous
